### PR TITLE
feat(dogstatsd): implement external data origin detection

### PR DESCRIPTION
**Pull request by wdhif**
### What does this PR do?

Implement External Data resolution for DogStatsD Origin Detection.

### Motivation

Follows the New Origin Detection Spec.

### Additional Notes

TODO Add profiling information

### Describe how to test/QA your changes

* Deploy the following workload:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: datadogpy
spec:
  replicas: 1
  selector:
    matchLabels:
      app: datadogpy
  template:
    metadata:
      labels:
        app: datadogpy
        admission.datadoghq.com/enabled: "true"
    spec:
      containers:
      - name: datadogpy
        image: wdhifdatadog/datadogpy
        imagePullPolicy: Always
        env:
        - name: USER
          value: "wdhif"
        - name: DD_DOGSTATSD_URL
          value: /var/run/datadog/dsd.socket
        - name: DD_AGENT_HOST
          valueFrom:
            fieldRef:
              fieldPath: status.hostIP
        - name: DD_ENTITY_ID
          valueFrom:
            fieldRef:
              fieldPath: metadata.uid
        volumeMounts:
        - mountPath: /var/run/datadog
          name: dogstatsd-socket
      volumes:
      - name: dogstatsd-socket
        hostPath:
          path: /var/run/datadog/
```

This modified workload is a modified version of [datadogpy](https://github.com/DataDog/datadogpy) that only send the External Data.

```
diff --git a/datadog/dogstatsd/base.py b/datadog/dogstatsd/base.py
index c284a74..734faee 100644
--- a/datadog/dogstatsd/base.py
+++ b/datadog/dogstatsd/base.py
@@ -995,14 +995,14 @@ class DogStatsd(object):
         self, metric, metric_type, value, tags, sample_rate=1, timestamp=0
     ):
         # Create/format the metric packet
-        return "%s%s:%s|%s%s%s%s%s%s" % (
+        return "%s%s:%s|%s%s%s%s%s" % (
             (self.namespace + ".") if self.namespace else "",
             metric,
             value,
             metric_type,
             ("|@" + text(sample_rate)) if sample_rate != 1 else "",
             ("|#" + ",".join(normalize_tags(tags))) if tags else "",
-            ("|c:" + self._container_id if self._container_id else ""),
+            #("|c:" + self._container_id if self._container_id else ""),
             ("|e:" + os.environ.get(EXTERNAL_DATA_ENV_VAR, "")),
             ("|T" + text(timestamp)) if timestamp > 0 else "",
         )
```

* Validate that you are still getting the metrics correctly tagged with the Container ID
image
image
image

### CODE REVIEW

**Review by gh123man — APPROVED** [review]

File: comp/core/tagger/taggerimpl/tagger.go [hunk 1]
```diff
@@ -46,6 +48,21 @@ import (
 
 var entityIDRegex = regexp.MustCompile(`^en-(init\.)?([a-fA-F0-9-]+)/([a-zA-Z0-9-_]+)$`)
 
+const (
+	// External Data Prefixes
+	// These prefixes are used to build the External Data Environment Variable.
+	// This variable is then used for Origin Detection.
+	externalDataInitPrefix          = "it-"
+	externalDataContainerNamePrefix = "cn-"
+	externalDataPodUIDPrefix        = "pu-"
```
Comment (clamoriniere) [inline comp/core/tagger/taggerimpl/tagger.go:57]: open question: Does it makes sense to have it publicly defined to be reuse by third party project like the DSD/APM libraries.

**Review by clamoriniere — APPROVED** [review]

File: comp/core/tagger/taggerimpl/tagger.go (hunk 1 above)
Comment (wdhif) [reply comp/core/tagger/taggerimpl/tagger.go:57]: Checked together, we will see how we do the implementation for Traces and improve this then if needed.

File: comp/core/tagger/taggerimpl/tagger_test.go [hunk 2]
```diff
@@ -153,3 +231,47 @@ func TestParseEntityID(t *testing.T) {
...
+		},
+		{
+			name:        "empty",
+			cardinality: "",
+			want:        tagger.DogstatsdCardinality(),
+		},
+		{
+			name:        "unknown",
+			cardinality: "foo",
+			want:        tagger.DogstatsdCardinality(),
```
[the remark this answers is not in the document]
Comment (wdhif) [reply comp/core/tagger/taggerimpl/tagger_test.go:269]: Checked together, we will add a comment in another PR.

**Pull request merged**